File: src/merge-sort.js
/**
 * Merge Sort is a Divide and Conquer algorithm, it divides a
 * large array into two smaller subarrays and then recursively
 * sort the subarrays.
 *
 * Complexity: O(nlog(n)).
 */
function mergeSort(arr) {
  if (arr.length <= 1) {
    return arr;
  }

  const mid = Math.floor(arr.length / 2);
  const left = arr.slice(0, mid);
  const right = arr.slice(mid);
  return merge(mergeSort(left), mergeSort(right));
}

function merge(left, right) {
  const result = [];

  while (left.length && right.length) {
    if (left[0] < right[0]) {
      result.push(left.shift());
    } else {
      result.push(right.shift());
    }
  }
  return [...result, ...left, ...right];
}

module.exports = mergeSort;

File: src/merge_sort.py
from dataclasses import dataclass
from typing import TypeVar, Generic

T = TypeVar('T', str, int)


@dataclass
class MergeSort(Generic[T]):
    array: list[T]

    def solve(self, array: list[T] = []) -> list[T]:
        if len(array) <= 1:
            return array
        middle = int(len(array) / 2)
        left = array[:middle]
        right = array[middle:]
        return self._merge(self.solve(left), self.solve(right))

    @staticmethod
    def _merge(left: list[T], right: list[T]) -> list[T]:
        result: list[T] = []
        while left and right:
            if left[0] < right[0]:
                result.append(left.pop(0))
                continue
            if left[0] >= right[0]:
                result.append(right.pop(0))
                continue
        return [*result, *left, *right]  # type: ignore
